File: example/android/app/src/main/java/app/loup/streams_channel_example/PluginExample.java
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

package app.loup.streams_channel_example;

import android.os.Handler;

import app.loup.streams_channel.StreamsChannel;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.PluginRegistry;

public class PluginExample implements FlutterPlugin {

  public static void registerWith(PluginRegistry.Registrar registrar) {
    final StreamsChannel channel = new StreamsChannel(registrar.messenger(), "streams_channel_example");
    channel.setStreamHandlerFactory(arguments -> new StreamHandler());
  }

  @Override
  public void onAttachedToEngine(FlutterPluginBinding binding) {
    final StreamsChannel channel = new StreamsChannel(binding.getBinaryMessenger(), "streams_channel_example");
    channel.setStreamHandlerFactory(arguments -> new StreamHandler());
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) {

  }

  // Send "Hello" 10 times, every second, then ends the stream
  public static class StreamHandler implements EventChannel.StreamHandler {

    private final Handler handler = new Handler();
    private final Runnable runnable = new Runnable() {
      @Override
      public void run() {
        if (count > 10) {
          eventSink.endOfStream();
        } else {
          eventSink.success("Hello " + count + "/10");
        }
        count++;
        handler.postDelayed(this, 1000);
      }
    };

    private EventChannel.EventSink eventSink;
    private int count = 1;

    @Override
    public void onListen(Object o, final EventChannel.EventSink eventSink) {
      System.out.println("StreamHandler - onListen: " + o);
      this.eventSink = eventSink;
      runnable.run();
    }

    @Override
    public void onCancel(Object o) {
      System.out.println("StreamHandler - onCancel: " + o);
      handler.removeCallbacks(runnable);
    }
  }
}

File: example/ios/Runner/PluginExample.m
//
//  PluginExample.m
//  Runner
//
//  Created by Lukasz on 2020-02-05.
//  Copyright © 2020 The Chromium Authors. All rights reserved.
//

#import "PluginExample.h"
#import <streams_channel/FlutterStreamsChannel.h>

@interface StreamHandler : NSObject<FlutterStreamHandler>
  @property(strong, nonatomic) NSTimer *timer;
  @property(assign, nonatomic) NSInteger count;
@end

@implementation PluginExample

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  
  FlutterStreamsChannel *channel = [FlutterStreamsChannel streamsChannelWithName:@"streams_channel_example" binaryMessenger:registrar.messenger];
  [channel setStreamHandlerFactory:^NSObject<FlutterStreamHandler> *(id arguments) {
    return [StreamHandler new];
  }];
}

@end

// Send "Hello" 10 times, every second, then ends the stream
@implementation StreamHandler

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events {
  NSLog(@"StreamHandler - onListen: %@", arguments);
  
  self.count = 1;
  self.timer = [NSTimer scheduledTimerWithTimeInterval:1 repeats:YES block:^(NSTimer * _Nonnull timer) {
    if(self.count > 10) {
      events(FlutterEndOfEventStream);
    } else {
      events([NSString stringWithFormat:@"Hello %ld/10", (long)self.count]);
      self.count++;
    }
  }];
  
  return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
  NSLog(@"StreamHandler - onCancel: %@", arguments);
  
  [self.timer invalidate];
  self.timer = nil;
  
  return nil;
}

@end
